Add CIP-62: Gingerbread Hardfork (#386)
celo-org/celo-proposals: CIPs/cip-0062.md

---
cip: 62
title: Gingerbread Hardfork
author: Paul Lange <@palango>
discussions-to: https://github.com/celo-org/celo-proposals/discussions/378
status: Final
type: Meta
created: 2023-06-20
license: Apache 2.0
---

This is the Meta CIP for the technical specifications of the Celo Gingerbread Hardfork.

## Simple Summary

Enable the activation of the Celo Gingerbread Hardfork specification which focuses on Ethereum alignment.

## Abstract

Add support for protocol-impacting changes that increase alignment with Ethereum by removing unused Celo-specific features and protocol differences. This hardfork is named _Gingerbread_ by the community.

This document proposes the following blocks at which to implement these changes in the Celo networks:

> - `18733000` on Baklava Testnet

Reviewer: ```suggestion
- `18785000` on Baklava Testnet
```

type: Meta
created: 2023-06-20
license: Apache 2.0
---

This is the Meta CIP for the technical specifications of the Celo Gingerbread Hardfork.

## Simple Summary

Enable the activation of the two parts of the Celo Gingerbread Hardfork specification which focuses on Ethereum alignment.

## Abstract

Add support for protocol-impacting changes that increase alignment with Ethereum by removing unused Celo-specific features and protocol differences. This hardfork is named _Gingerbread_ by the community.

This document proposes the following blocks at which to implement these changes in the Celo networks:

- Baklava Testnet:
  - Part 1: on block `18785000`
  - Part 2: on block `TODO`
- Alfajores Testnet, both Parts on block `TODO`
- Celo Mainnet, both Parts on block `TODO`

For more information on the opcodes and their respective EIP and CIP implementations, please see the _Specification_ section of this document.

## Motivation

To enhance the EVM's capabilities and while pushing forward the Celo blockchain functionality in order to enable economic conditions for prosperity as well as reducing differences with Ethereum which will lead to increased compatibility and ease maintenance.

> A second Part of the Gingerbread Hardfork was decided later to reduce even more the differences with Ethereum, and give us some future tools required for Celo's switch to became a Layer 2.

Reviewer 2: ```suggestion
A second Part of the Gingerbread Hardfork was decided later to reduce even more the differences with Ethereum, and give us some future tools required for Celo's switch to become a Layer 2.
```

- [CIP-58: Restore opcode gas costs from Ethereum](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0058.md)
- [CIP-59: Re-add header fields with values used in Ethereum 2.0](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0059.md)
- [CIP-60: Add GASLIMIT opcode](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0060.md)
- [CIP-61: Add BASEFEE opcode](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0061.md)

CIPs to be included in the Part 2:

- [CIP-63: Remove BLS-12377, BLS-12381, CIP-20 and CIP-25 precompiles](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0063.md)
- [CIP-64: New Transaction Type: Celo Dynamic Fee v2](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0064.md)
- [CIP-65: Total Transaction Data Limit](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0065.md)

With the above CIPs the block header contains the same data as the Ethereum chain. Therefore, the hash of the header will be changed so that it is compatible with Ethereum.

### Deprecation Warning

> As part of Celo's switch to became a Layer 2 and the deprecation of the full node incentives (part of this Hard Fork CIP-57) in future Hard Forks:

Reviewer 2: ```suggestion
As part of Celo's switch to become a Layer 2 and the deprecation of the full node incentives (part of this Hard Fork CIP-57) in future Hard Forks:
```

- The Celo Legacy Transaction: `rlp(nonce, gasPrice, gas, feeCurrency, gatewayFeeRecipient, gatewayFee, to, value, data, signatureV, signatureR, signatureS)` will be changed to the original Ethereum transaction `rlp(nonce, gasPrice, gas, to, value, data, signatureV, signatureR, signatureS)`, removing not only the gateway fields, but also the `feeCurrency`.
- The Enveloped Transaction created in the [CIP-42](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0042.md) (Type: `0x7c`) will be completely deprecated

As the full node incentives are removed these transactions not only consume unnecesary space, but bring unnecessary complexity to our codebase.

> For the sake of maintaining the functionality of paying fees in any of our whitelisted currencies the [CIP-64](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0064.md) was added to this Hard Fork to have enough time to migrate any functionality that uses the old way to pay for transactions in differenc.
> 
> So, in resume, 

Reviewer 2: This looks like something is missing at the end. Suggestion:

```suggestion
For the sake of maintaining the functionality of paying fees in any of our whitelisted currencies the [CIP-64](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0064.md) was added to this Hard Fork. The time until CIP-42 transactions are removed can be used to migrate to CIP-64 transactions.
```
